src/IB/IBFESurfaceMethod.cpp

```diff
@@ -1388,162 +2289,1758 @@ void IBFESurfaceMethod::endDataRedistribution(Pointer<PatchHierarchy<NDIM> > /*h
+                {
+                    const double p_JxW = phi[k][qp] * JxW[qp];
+                    P_o_rhs_e(k) += P_o_qp[idx] * p_JxW;
+                    P_i_rhs_e(k) += P_i_qp[idx] * p_JxW;
+                }
+            }
+
+            P_o_dof_map.constrain_element_vector(P_o_rhs_e, P_o_dof_indices);
+            P_o_rhs_vec->add_vector(P_o_rhs_e, P_o_dof_indices);
+            
+            P_i_dof_map.constrain_element_vector(P_i_rhs_e, P_i_dof_indices);
+            P_i_rhs_vec->add_vector(P_i_rhs_e, P_i_dof_indices);
+            
+            qp_offset += n_qp;
+        }
+    }
+    P_o_rhs_vec->close();
+	P_i_rhs_vec->close();
+	
+    d_fe_data_managers[part]->computeL2Projection(
+        *P_o_vec, *P_o_rhs_vec, P_O_SYSTEM_NAME, d_default_interp_spec.use_consistent_mass_matrix);
+
+    d_fe_data_managers[part]->computeL2Projection(
+        *P_i_vec, *P_i_rhs_vec, P_I_SYSTEM_NAME, d_default_interp_spec.use_consistent_mass_matrix);
+
+    d_X_half_vecs[part]->close();
+    d_X_current_vecs[part]->close();
+    d_X_new_vecs[part]->close();
+
+    d_P_o_half_vecs[part]->close();
+    d_P_i_half_vecs[part]->close();
+    d_P_j_IB_ghost_vecs[part]->close();
+    d_X_IB_ghost_vecs[part]->close();
+
+    return;
+
+} // interpolatePressureForTraction
+
+void
+IBFESurfaceMethod::computeFluidTraction(const double data_time, unsigned int part)
+{
+    NumericVector<double>* WSS_o_vec = NULL;
+    NumericVector<double>* WSS_o_ghost_vec = d_WSS_o_IB_ghost_vecs[part];
+
+    NumericVector<double>* P_o_vec = NULL;
+    NumericVector<double>* P_o_ghost_vec = d_P_o_IB_ghost_vecs[part];
+
+    NumericVector<double>* P_j_vec = NULL;
+    NumericVector<double>* P_j_ghost_vec = d_P_j_IB_ghost_vecs[part];
+
+    boost::array<PetscVector<double>*, NDIM> DU_j_ghost_vec;
+    boost::array<PetscVector<double>*, NDIM> DU_j_vec;
+
+    NumericVector<double>* TAU_vec = d_TAU_half_vecs[part];
+
+    NumericVector<double>* X_vec = NULL;
+
+    if (MathUtilities<double>::equalEps(data_time, d_current_time))
+    {
+        X_vec = d_X_current_vecs[part];
+    }
+    else if (MathUtilities<double>::equalEps(data_time, d_half_time))
+    {
+        X_vec = d_X_half_vecs[part];
+    }
+    else if (MathUtilities<double>::equalEps(data_time, d_new_time))
+    {
+        X_vec = d_X_new_vecs[part];
+    }
+    NumericVector<double>* X_ghost_vec = d_X_IB_ghost_vecs[part];
+    X_vec->localize(*X_ghost_vec);
+    
+    WSS_o_vec = d_WSS_o_half_vecs[part];
+    WSS_o_vec->localize(*WSS_o_ghost_vec);
+
+    P_o_vec = d_P_o_half_vecs[part];
+    P_o_vec->localize(*P_o_ghost_vec);
+
+    P_j_vec = d_P_j_half_vecs[part];
+    P_j_vec->localize(*P_j_ghost_vec);
+
+    DU_j_ghost_vec = d_DU_j_IB_ghost_vecs[part];
+    DU_j_vec = d_DU_j_half_vecs[part];
+    for (unsigned int d = 0; d < NDIM; ++d)
+    {
+        DU_j_vec[d]->localize(*DU_j_ghost_vec[d]);
+    }
+    for (unsigned int d = 0; d < NDIM; ++d)
+    {
+        DU_j_vec[d]->localize(*DU_j_ghost_vec[d]);
+    }
```

Comment:
Like elsewhere, localization should not be necessary. Explicitly update the ghost region.